[kbn-config-schema] Fix error in `schema.literal` type

nickofthyme opened the pull request:
## Summary

When using `@kbn/config-schema` with the `literal` type, it works fine in every case except for when the `value` passed to the validator is `undefined`.

```ts
schema.literal(1).validate(undefined)
```

The resulting error is actually a runtime error as it tries to destructure `valids`.

```
Cannot destructure property 'Symbol(Symbol.iterator)' of 'undefined' as it is undefined.
```

https://github.com/elastic/kibana/blob/67a76e46b45db96b04c0042aa9383285c7e06284/src/platform/packages/shared/kbn-config-schema/src/types/literal_type.ts#L18-L24

The problem is that `handleError` above is called with some `context`, but this `context` for the case above, is missing `valids` as you can see in the debug state below.

![Zight 2025-07-24 at 4 06 02 PM](https://github.com/user-attachments/assets/52dfc4fe-1c0d-4683-baf0-1392b782acc9)

The solution I propose is to properly handle the missing `valids` and store the `expectedValue` on the `LiteralType` class. This way we can fallback to this value when the expected value is not contained in the context. This type is unique to all other as it is expecting an exact value not a general value of some type, so I think think this solution is ideal. I also tried looking for the expected value within the base `Type` instance but couldn't find anything.

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

<!--ONMERGE {"backportTargets":["9.1"]} ONMERGE-->

nickofthyme committed:
fix: error in `@kbn/config-schema` literal type

nickofthyme committed:
test: add test for use with maybe

nickofthyme committed:
Merge branch 'main' into fix-literal-type-error

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/jest_integration -u src/core/server/integration_tests/ci_checks'

jloleysens reviewed (commented):
[Scratch my original comment -- re-read!]

---

Side note/my 2c: changing validation messages can affect a lot of snapshots :(... Might want to avoid that update for now

jloleysens commented on src/platform/packages/shared/kbn-config-schema/src/types/literal_type.ts:
```diff
@@ -11,15 +11,26 @@ import { internals } from '../internals';
   }
 
-  protected handleError(type: string, { value, valids: [expectedValue] }: Record<string, any>) {
+  protected handleError(
+    type: string,
+    { value, valids: [expectedValue = this.expectedValue] = [] }: Record<string, any>
+  ) {
+    const formatter = (v: any) =>
```
WDYT about avoid this "valids" array entirely. Something like:

```ts
export class LiteralType<T> extends Type<T> {
  private expectedValue: T;
  constructor(value: T) {
    super(internals.any().valid(value));
    this.expectedValue = value;
  }

  protected handleError(type: string) {
    switch (type) {
      case 'any.required':
      case 'any.only':
        return `expected value to equal [${this.expectedValue}]`;
    }
  }
}
```

jloleysens replied:
Also would be nice to remove these `any`s :)

nickofthyme replied:
Yeah good ideas. Applied all changes in 5a719d9.

I also removed the changes to the error message, partly I wanted to see how many failures it would cause. Nope, not worth it 😅.

nickofthyme commented on src/core/server/integration_tests/ci_checks/saved_objects/check_registered_types.test.ts:
Yeah no clue what that was about.

nickofthyme committed:
chore: update per comments

- revert error message change
- remove `valids` in favor of `expectedValue`
- update any time to generic

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/323452)
* Commit: 5a719d9079032294d4646caf67459e9c246aa93f

### Failed CI Steps
* [Jest Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/323452#0198425d-4532-4124-b599-62550a5d254d)
* [Jest Tests #5](https://buildkite.com/elastic/kibana-pull-request/builds/323452#0198425d-4533-46a5-bd42-04b085965d4c)
* [Jest Tests #8](https://buildkite.com/elastic/kibana-pull-request/builds/323452#0198425d-4534-455a-a75e-7525b3114f5e)
* [Jest Tests #8](https://buildkite.com/elastic/kibana-pull-request/builds/323452#01984281-7822-4772-af4f-849156c931f5)
* [Jest Tests #11](https://buildkite.com/elastic/kibana-pull-request/builds/323452#0198425d-4535-4000-9fad-84e9642eebdb)
* [Jest Tests #14](https://buildkite.com/elastic/kibana-pull-request/builds/323452#0198425d-4536-498b-8e1f-f69b452a5883)
* [FTR Configs #68](https://buildkite.com/elastic/kibana-pull-request/builds/323452#0198425d-45b4-4e04-a4af-aaecfbd0424a)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323452#0198425d-45b4-4e04-a4af-aaecfbd0424a) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323452/jobs/0198425d-45b4-4e04-a4af-aaecfbd0424a/artifacts/01984288-b671-45e9-be95-124b0f4ad1a1) FTR Configs #<span></span>68 / cloud_security_posture Serverless - Agentless CIS Integration Page Serverless - Agentless CIS_AWS Create flow user should save agentless integration policy when there are no api or validation errors and button is not disabled
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323452#01984281-7822-4772-af4f-849156c931f5) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323452/jobs/01984281-7822-4772-af4f-849156c931f5/artifacts/01984294-46d2-437c-bbe9-5f0f6f8c1717) Jest Tests #<span></span>8 / validate params should validate and throw error when params are invalid

### History
* :broken_heart: [Build #323085](https://buildkite.com/elastic/kibana-pull-request/builds/323085) failed 70133591785ebaeb16989db862c74562fc46dc41

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323452","number":323452,"commit":"5a719d9079032294d4646caf67459e9c246aa93f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323085","number":323085,"commit":"70133591785ebaeb16989db862c74562fc46dc41"}],"number":323452}
buildkite-pr-comment-->

nickofthyme committed:
test: update error messages

nickofthyme committed:
Merge branch 'main' into fix-literal-type-error

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/323514)
* Commit: 74aa65bff718014793a292faa995dc21f3d13a7d

### Failed CI Steps
* [Jest Tests #2](https://buildkite.com/elastic/kibana-pull-request/builds/323514#019842e8-039e-4518-bf19-daa4e57e1850)
* [Jest Tests #4](https://buildkite.com/elastic/kibana-pull-request/builds/323514#019842e8-039f-46a0-bd55-27db6e6e4e1a)
* [Jest Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/323514#019842e8-039f-4feb-850d-55fffe7f9e61)
* [Jest Tests #10](https://buildkite.com/elastic/kibana-pull-request/builds/323514#019842e8-03a0-4baa-bc6b-9dbe48d47f7a)
* [Jest Tests #13](https://buildkite.com/elastic/kibana-pull-request/builds/323514#019842e8-03a0-44f7-ac23-268f72b3e70c)
* [Jest Tests #16](https://buildkite.com/elastic/kibana-pull-request/builds/323514#019842e8-03a1-4f5f-b0eb-60c110f0e151)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/67a76e46b45db96b04c0042aa9383285c7e06284) | [after](https://github.com/elastic/kibana/commit/74aa65bff718014793a292faa995dc21f3d13a7d) | diff |
| --- | --- | --- | --- |
| `dataUsage` | 244.3KB | 244.4KB | +60.0B |



### History
* :broken_heart: [Build #323452](https://buildkite.com/elastic/kibana-pull-request/builds/323452) failed 5a719d9079032294d4646caf67459e9c246aa93f
* :broken_heart: [Build #323085](https://buildkite.com/elastic/kibana-pull-request/builds/323085) failed 70133591785ebaeb16989db862c74562fc46dc41

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323514","number":323514,"commit":"74aa65bff718014793a292faa995dc21f3d13a7d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323452","number":323452,"commit":"5a719d9079032294d4646caf67459e9c246aa93f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323085","number":323085,"commit":"70133591785ebaeb16989db862c74562fc46dc41"}],"number":323514}
buildkite-pr-comment-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/323515)
* Commit: e4d3b82413a04d629a62cf515535d4c8ff238f4d

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/2c82a71bf3da1070351f85f8d5860797dcdf9f2c) | [after](https://github.com/elastic/kibana/commit/e4d3b82413a04d629a62cf515535d4c8ff238f4d) | diff |
| --- | --- | --- | --- |
| `dataUsage` | 244.3KB | 244.4KB | +60.0B |



### History
* :yellow_heart: [Build #323514](https://buildkite.com/elastic/kibana-pull-request/builds/323514) was flaky 74aa65bff718014793a292faa995dc21f3d13a7d
* :broken_heart: [Build #323452](https://buildkite.com/elastic/kibana-pull-request/builds/323452) failed 5a719d9079032294d4646caf67459e9c246aa93f
* :broken_heart: [Build #323085](https://buildkite.com/elastic/kibana-pull-request/builds/323085) failed 70133591785ebaeb16989db862c74562fc46dc41

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323515","number":323515,"commit":"e4d3b82413a04d629a62cf515535d4c8ff238f4d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323514","number":323514,"commit":"74aa65bff718014793a292faa995dc21f3d13a7d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323452","number":323452,"commit":"5a719d9079032294d4646caf67459e9c246aa93f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323085","number":323085,"commit":"70133591785ebaeb16989db862c74562fc46dc41"}],"number":323515}
buildkite-pr-comment-->

nickofthyme committed:
revert automated changes

nickofthyme committed:
Merge branch 'main' into fix-literal-type-error

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/jest_integration -u src/core/server/integration_tests/ci_checks'

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/323578)
* Commit: 72d4cb146fbdfb9a50d1a290b497de276778723a

### Failed CI Steps
* [FTR Configs #83](https://buildkite.com/elastic/kibana-pull-request/builds/323578#0198481e-ebe1-43f8-a51d-3fd60f324adf)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/323578#0198481e-ebe1-43f8-a51d-3fd60f324adf) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/323578/jobs/0198481e-ebe1-43f8-a51d-3fd60f324adf/artifacts/01984843-0d06-4f2b-bce1-55c4e85e6601) FTR Configs #<span></span>83 / Air-gapped environment with pre-bundled packages @<span></span>ess @<span></span>serverless @<span></span>skipInServerlessMKI Bootstrap Prebuilt Rules installs required Fleet packages required for detection engine to function

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/4b1de132de04220d47bb71de9c59171b167d886b) | [after](https://github.com/elastic/kibana/commit/72d4cb146fbdfb9a50d1a290b497de276778723a) | diff |
| --- | --- | --- | --- |
| `dataUsage` | 244.3KB | 244.4KB | +60.0B |



### History
* :green_heart: [Build #323515](https://buildkite.com/elastic/kibana-pull-request/builds/323515) succeeded e4d3b82413a04d629a62cf515535d4c8ff238f4d
* :yellow_heart: [Build #323514](https://buildkite.com/elastic/kibana-pull-request/builds/323514) was flaky 74aa65bff718014793a292faa995dc21f3d13a7d
* :broken_heart: [Build #323452](https://buildkite.com/elastic/kibana-pull-request/builds/323452) failed 5a719d9079032294d4646caf67459e9c246aa93f
* :broken_heart: [Build #323085](https://buildkite.com/elastic/kibana-pull-request/builds/323085) failed 70133591785ebaeb16989db862c74562fc46dc41

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323578","number":323578,"commit":"72d4cb146fbdfb9a50d1a290b497de276778723a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323515","number":323515,"commit":"e4d3b82413a04d629a62cf515535d4c8ff238f4d"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323514","number":323514,"commit":"74aa65bff718014793a292faa995dc21f3d13a7d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323452","number":323452,"commit":"5a719d9079032294d4646caf67459e9c246aa93f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323085","number":323085,"commit":"70133591785ebaeb16989db862c74562fc46dc41"}],"number":323578}
buildkite-pr-comment-->

jcger reviewed (approved):
thanks!

jloleysens reviewed (approved):
Thanks for addressing my feedback!

kibanamachine commented:
Starting backport for target branches: 9.1

https://github.com/elastic/kibana/actions/runs/16576054319
